wire-kotlin-generator/src/main/java/com/squareup/wire/kotlin/KotlinGenerator.kt

```diff
@@ -0,0 +1,399 @@
+          fieldClass = fieldClass.asNullable()
+        }
+        field.isRepeated -> {
+          fieldClass = List::class.asClassName().parameterizedBy(fieldClass)
+        }
+      }
+
+      val parameterSpec = ParameterSpec.builder(fieldName, fieldClass)
+      if (field.isOptional) {
+        parameterSpec.defaultValue("null")
+      }
+      constructorBuilder.addParameter(parameterSpec.build())
+      classBuilder.addProperty(PropertySpec.builder(fieldName, fieldClass)
+          .initializer(fieldName)
+          .build())
+    }
+
+    val unknownFields = nameAllocator.get("unknownFields")
+    constructorBuilder.addParameter(
+        ParameterSpec.builder(unknownFields, ByteString::class.asClassName())
+            .defaultValue("%T.EMPTY", ByteString::class.asClassName())
+            .build())
+    classBuilder.addProperty(PropertySpec.builder(unknownFields, ByteString::class.asClassName())
+        .initializer(unknownFields)
+        .build())
+
+    classBuilder.primaryConstructor(constructorBuilder.build())
+  }
+
+  private fun encodedSizeFunc(message: MessageType): FunSpec {
+    val className = generatedTypeName(message)
+    val body = CodeBlock.builder()
+        .add("return ")
+
+    message.fields().forEach { field ->
+      val adapterName = getAdapterName(field)
+
+      body.add("%L.%LencodedSizeWithTag(%L, value.%L) + \n",
+          adapterName, if (field.isRepeated) "asRepeated()." else "", field.tag(), field.name())
+    }
+
+    return FunSpec.builder("encodedSize")
+        .addParameter("value", className)
+        .returns(Int::class.asTypeName())
+        .addCode(body
+            .add("value.unknownFields.size()\n")
+            .build())
+        .addModifiers(OVERRIDE)
+        .build()
+  }
+
+  private fun encodeFunc(message: MessageType): FunSpec {
+    val className = generatedTypeName(message)
+    var body = CodeBlock.builder()
+
+    message.fields().forEach { field ->
+      val adapterName = getAdapterName(field)
+      body.addStatement("%L.%LencodeWithTag(writer, %L, value.%L)",
+          adapterName, if (field.isRepeated) "asRepeated()." else "", field.tag(), field.name())
+    }
+
+    body.addStatement("writer.writeBytes(value.unknownFields)")
+
+    return FunSpec.builder("encode")
+        .addParameter("writer", ProtoWriter::class.asClassName())
+        .addParameter("value", className)
+        .addCode(body.build())
+        .addModifiers(OVERRIDE)
+        .build()
+  }
+
+  private fun decodeFunc(message: MessageType): FunSpec {
+    val className = nameToKotlinName[message.type()]!!
+
+    var declarationBody = CodeBlock.builder()
+
+    var returnBody = CodeBlock.builder()
+    returnBody.add("return %L(\n", className.simpleName)
+
+    var decodeBlock = CodeBlock.builder()
+    decodeBlock.addStatement("val unknownFields = reader.decodeMessage { tag -> ")
+
+    val INDENTATION = " ".repeat(4)
+
+    // Indent manually as code generator doesn't handle this block gracefully.
```

Comment:
File a KotlinPoet bug? I'm surprised you can't use begin|endControlFlow like

https://github.com/square/kotlinpoet/blob/aa22394acdcb087f413ecb6561980573460e7877/src/test/java/com/squareup/kotlinpoet/FunSpecTest.kt#L376-L379

Reply:
I am not sure how to make kotlin poet generate this statement 

`val unknownFields = reader.decodeMessage { tag ->`

the full code would be

```
     val unknownFields = reader.decodeMessage { tag ->
                when (tag) {
                    1 -> name = ProtoAdapter.STRING.decode(reader)
                    2 -> id = ProtoAdapter.INT32.decode(reader)
                    3 -> email = ProtoAdapter.STRING.decode(reader)
                    4 -> phone.add(PhoneNumber.ADAPTER.decode(reader))
                    else -> UnkownFieldsBuilder.UNKNOWN_FIELD
                }
            }
```

Reply:
I like the way1 better, taking it over here https://github.com/square/kotlinpoet/issues/416